carp.common/src/commonMain/kotlin/dk/cachet/carp/common/users/ParticipantAttribute.kt

```diff
@@ -55,6 +54,26 @@ sealed class ParticipantAttribute
             is CustomParticipantAttribute<*> -> input
         }
 
+    /**
+     * Determines whether [input] is valid and can be converted to a matching [Data] object associated to this attribute
+     * as registered in [registeredInputDataTypes] or [CustomInput] in case this is a [CustomParticipantAttribute].
+     *
+     * @throws UnsupportedOperationException when no input element is registered for this attribute.
+     */
+    fun <TInput> isValidInput( registeredInputDataTypes: InputDataTypeList, input: TInput ): Boolean
+    {
+        @Suppress( "UNCHECKED_CAST" )
```

Comment:
Is it possible to avoid these by using the kotlin safe cast `as?` ?

Reply:
No. This has to do with limitations of generics on JVM. :( I could _not_ do the cast, in which case it would be `InputElement<*>` which can be inferred automatically.

But, then a cast is needed of `input` at:
`return isExpectedDataType && inputElement.isValid( input )`

This is thus working around generics limitations of the JVM runtime by using `inputElement.getDataClass().isInstance( input )` to verify whether or not the types match. `getDataClass` wouldn't be needed with true generics.